// fela-monolithic/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = monolithic;

var _isPlainObject = require('lodash/isPlainObject');

var _isPlainObject2 = _interopRequireDefault(_isPlainObject);

var _reduce = require('lodash/reduce');

var _reduce2 = _interopRequireDefault(_reduce);

var _cssifyObject = require('css-in-js-utils/lib/cssifyObject');

var _cssifyObject2 = _interopRequireDefault(_cssifyObject);

var _felaUtils = require('fela-utils');

var _generateMonolithicClassName = require('./generateMonolithicClassName');

var _generateMonolithicClassName2 = _interopRequireDefault(_generateMonolithicClassName);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function useMonolithicRenderer(renderer) {
  var prettySelectors = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

  renderer.prettySelectors = prettySelectors;

  // monolithic output can not be rehydrated
  renderer.enableRehydration = false;

  renderer._renderStyleToCache = function (className, style) {
    var pseudo = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '';
    var media = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '';
    var support = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : '';

    var ruleSet = (0, _reduce2.default)(style, function (ruleset, value, property) {
      if ((0, _isPlainObject2.default)(value)) {
        if ((0, _felaUtils.isNestedSelector)(property)) {
          renderer._renderStyleToCache(className, value, pseudo + (0, _felaUtils.normalizeNestedProperty)(property), media, support);
        } else if ((0, _felaUtils.isMediaQuery)(property)) {
          var combinedMediaQuery = (0, _felaUtils.generateCombinedMediaQuery)(media, property.slice(6).trim());

          renderer._renderStyleToCache(className, value, pseudo, combinedMediaQuery, support);
        } else if ((0, _felaUtils.isSupport)(property)) {
          var combinedSupport = (0, _felaUtils.generateCombinedMediaQuery)(support, property.slice(9).trim());
          renderer._renderStyleToCache(className, value, pseudo, media, combinedSupport);
        } else {
          // TODO: warning
        }
      } else if (!(0, _felaUtils.isUndefinedValue)(value)) {
        ruleset[property] = value;
      }

      return ruleset;
    }, {});

    if (Object.keys(ruleSet).length > 0) {
      var css = (0, _cssifyObject2.default)(ruleSet);
      var selector = (0, _felaUtils.generateCSSSelector)(className, pseudo);

      var change = {
        type: _felaUtils.RULE_TYPE,
        className: className,
        selector: selector,
        declaration: css,
        media: media
      };

      var declarationReference = selector + media + support;
      renderer.cache[declarationReference] = change;
      renderer._emitChange(change);
    }
  };

  renderer._renderStyleToClassNames = function (style, rule) {
    if (Object.keys(style).length < 1) {
      return '';
    }

    var localRulePrefix = renderer.prettySelectors && (rule.ruleName || rule.name) ? (rule.ruleName || rule.name) + '_' : '';

    var className = (0, _generateMonolithicClassName2.default)(style, (renderer.selectorPrefix || '') + (rule.selectorPrefix || localRulePrefix));

    if (!renderer.cache.hasOwnProperty(className)) {
      renderer._renderStyleToCache(className, style);
      renderer.cache[className] = {};
    }

    return className;
  };

  renderer.renderRule = function (rule) {
    var props = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    var processedStyle = (0, _felaUtils.processStyleWithPlugins)(renderer, rule(props, renderer), _felaUtils.RULE_TYPE, props);
    return renderer._renderStyleToClassNames(processedStyle, rule);
  };

  return renderer;
}
/* eslint-disable no-continue */
function monolithic() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  return function (renderer) {
    return useMonolithicRenderer(renderer, process.env.NODE_ENV !== 'production' && options.prettySelectors);
  };
}
